### PR TITLE
Additional parameterisation

### PR DESCRIPTION
**Pull request by thomashaw**
### List of new modules:
	
 - vulnerabilities/unix/system/parameterised_accounts -- still a work in process but usable!
		Has one fact: accounts[] which is an array of Account Hashes.
		An account hash is created using the account_hash_builder encoder.
 - encoders/account/account_hash_builder -- This enables us to encapsulate accounts into individual hashes so we can have different random inputs on each. Takes input of a username, password, super_user(boolean), strings_to_leak[] and leaked_filename and outputs the values in a hash.
 - generators/random/random_boolean -- returns a random true / false
 - generators/random/random_common_password -- top 100 passwords from wordlist
 - generators/random/random_weak_password -- random 4-6 character name/noun from wordlist
 - generators/random/random_medium_password -- random 6 letter word + 2 digit number
 - generators/random/random_strong_password -- random 15char base64 string
 - generators/random/random_username  -- random adjective +['_' '-', ''].sample + noun
 - generators/messages/welcome_message -- selects 1 of 3 welcome messages
 - generators/filenames/leaked_filename -- outputs random filename from a list
 - generators/filenames/storage_directory -- outputs random generated storage directory path 

### List of facts added: 
 - proftpd_133c_backdoor: server_name, welcome_msg
 - vsftpd_234_backdoor: anonymous_ftp, ftpd_banner
 - chkrootkit: cron_frequency,
 - distcc_exec: strings_to_leak, leaked_filenames
 - nfs_overshare: leaked_filename, storage_directory
 - gitlist_040: strings_to_leak, leaked_filename, github_repository
 - moinmoin_195: strings_to_leak, site_name, default_page
 - unrealirc_2381_backdoor: strings_to_leak, leaked_filename, server_name, server_desc, 
						admin_name, admin_email, user, group, motd 

### Misc changes 
 - generator/random_word: Removed comments
 - service/apache: changes for parameterisation
 - service/samba: changes for parameterisation
 - service/unrealirc: changes for parameterisation
 - builds: Based on the Utilities schema. I wrote the framework for this as i was going to use a puppetlabs sudo module but decided to manually handle that. Left th	builds: Based on the Utilities schema. I wrote the framework for this as i was going to use ais in for when I update the cleanup module.
 - unix/system/accounts: (upgraded to latest version) 
 - unix_update: removed conflict with windows as it conficts w/ generators+encoders that specify both linux and windows: when windows integration comes in we will handle platform clashes at the selection stage)
 - scenarios/parameterised_examples/parameterised_accounts.xml new scenario

### Next actions

Add strings_to_leak to: 
 - local/chkrootkit, 
 - local/setuid_nmap
 - proftpd_133c_backdoor
 - vsftpd backdoor 
 - shellshock_apache_cgi
 - writable_shadow?
 - nfs_rootshare

Finish off:
 - storage_directory generator's filler_path + nfs / smb modules handling of creating multiple directories at a time (see: http://stackoverflow.com/questions/21882525/puppet-recursive-directories-creation).

### Future considerations
We want to be able to cache the list of generated users + pass them into the username fields of scenarios? It would be cool to have consistency across all configurations on the same system. 
Would be nice to pick the server_name from the <server_name> in the scenario and feed that in as a default_input. This needs a bit more thought.

### CODE REVIEW

File: modules/encoders/account/account_hash_builder/secgen_metadata.xml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+  <read_fact>username</read_fact>
+  <read_fact>password</read_fact>
+  <read_fact>super_user</read_fact>
+  <read_fact>strings_to_leak</read_fact>
+  <read_fact>leaked_filename</read_fact>
+
+  <!-- TODO: This encoder should have default values so that we can omit a default input -->
+
+  <write_fact>account</write_fact>
```
Comment (cliffe) [inline modules/encoders/account/account_hash_builder/secgen_metadata.xml]: I think we should change "write_fact" to "output_type", at some point -- which would be more helpful for filtering rules -- "write_fact" is not actively used now

File: modules/generators/random/random_word/secgen_local/local.rb [hunk 2]
```diff
@@ -8,16 +8,7 @@ def initialize
...
-    #
-    # list = Wordlist::FlatFile.new("#{ROOT_DIR}/lib/resources/wordlists/wordlist")
-    # list.each_unique do |word|
-    #   outputs << word
-    #   break
-    # end
-
-    self.outputs << File.readlines("#{ROOT_DIR}/lib/resources/wordlists/wordlist").sample.chomp
-
+    self.outputs << File.readlines("#{WORDLISTS_PATH}/wordlist").sample.chomp
```
Comment (cliffe) [inline modules/generators/random/random_word/secgen_local/local.rb]: This is better. But can we call it WORDLISTS_DIR to be more consistent? (Possibly updating any other "PATH" constants also)

File: modules/vulnerabilities/unix/irc/unrealirc_3281_backdoor/templates/overshare.erb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+<% require 'json' -%>
+<%= JSON.parse(@json_inputs)["strings_to_leak"].join("\n----\n") %>
```
Comment (cliffe) [inline modules/vulnerabilities/unix/irc/unrealirc_3281_backdoor/templates/overshare.erb]: There is no reason the strings can't all end up in separate files -- the NFS example just dumped them into a file as a simple example

File: modules/vulnerabilities/unix/local/writeable_shadow/secgen_metadata.xml [hunk 4]
```diff
@@ -23,9 +23,4 @@
   <hint>An access control misconfiguration</hint>
   <solution>Edit the shadow file to set a password for root</solution>
 
-  <conflict>
```
Comment (cliffe) [inline modules/vulnerabilities/unix/local/writeable_shadow/secgen_metadata.xml]: Not sure why this has been removed. It shouldn't select this vuln twice for a system

File: modules/vulnerabilities/unix/nfs/nfs_overshare/templates/exports.erb [hunk 5]
```diff
@@ -1,9 +1,12 @@
+<%require 'json' -%>
+<%$parsed_inputs = JSON.parse(@json_inputs) -%>
+<%$storage_directory = $parsed_inputs['storage_directory'].first -%>
 # /etc/exports: the access control list for filesystems which may be exported
 #               to NFS clients.  See exports(5).
 #
 # Example for NFSv2 and NFSv3:
- /exports       172.0.0.0/8(rw,sync,no_subtree_check)
- /exports       192.0.0.0/8(rw,sync,no_subtree_check)
+ <%=$storage_directory %>       172.0.0.0/8(rw,sync,no_subtree_check)
```
Comment (cliffe) [inline modules/vulnerabilities/unix/nfs/nfs_overshare/templates/exports.erb]: I think we should be using 0.0.0.0 so we are not tying ourselves to those two network ranges

File: modules/vulnerabilities/unix/webapp/moinmoin_195/secgen_metadata.xml [hunk 6]
```diff
@@ -17,6 +17,23 @@
   <access>remote</access>
   <platform>linux</platform>
 
+  <read_fact>strings_to_leak</read_fact>
+  <read_fact>site_name</read_fact>
+  <read_fact>default_page</read_fact>
+
+  <default_input into="strings_to_leak">
+    <value>This is some leaked data.</value>
+    <generator type="string_generator"/>
```
Comment (cliffe) [inline modules/vulnerabilities/unix/webapp/moinmoin_195/secgen_metadata.xml]: I think it helps to be more specific that, as you may end up with a random string that they could spend hours trying to crack, better to generate messages, etc

File: modules/vulnerabilities/unix/webapp/moinmoin_195/secgen_metadata.xml [hunk 7]
```diff
@@ -17,6 +17,23 @@
...
+  </default_input>
+
+  <default_input into="site_name">
+    <value>Example site name!</value>
```
Comment (cliffe) [inline modules/vulnerabilities/unix/webapp/moinmoin_195/secgen_metadata.xml]: Can we generate one?

File: modules/vulnerabilities/unix/nfs/nfs_overshare/templates/exports.erb (hunk 5 above)
Comment (cliffe) [inline modules/vulnerabilities/unix/nfs/nfs_overshare/templates/exports.erb]: (not a regression, so could leave for a separate PR)

**Comment by cliffe** [comment]
I suggest we move helloworld to the messages directory

**Comment by thomashaw** [comment]
Thanks for the feedback. I've made and tested the suggested changes so I'm going to merge this into master. Cheers!
